server/src/main/java/org/jboss/as/server/deployment/DeploymentUnitPhaseService.java

```diff
@@ -126,7 +126,8 @@ public void transition(final ServiceController<?> controller, final ServiceContr
         if (nextPhase != null) {
             final ServiceName serviceName = DeploymentUtils.getDeploymentUnitPhaseServiceName(deploymentUnit, nextPhase);
             phaseService = DeploymentUnitPhaseService.create(deploymentUnit, nextPhase);
-            phaseServiceBuilder = serviceTarget.addService(serviceName, phaseService);
+            DeploymentUnitPhaseBuilder builder = this.deploymentUnit.getAttachment(Attachments.DEPLOYMENT_UNIT_PHASE_BUILDER);
```

Comment:
So this happens before the deployers for the current phase are run. So if you attach this in a given phase it won't affect the next phase, but the one after. Also because it is a DeploymentUnit attachment and not a phase context attachment it will then affect every phase from then on. Is this the behaviour you really want?